tests/console/commands/test_run.py

```diff
@@ -146,3 +148,38 @@ def test_run_script_exit_code(
     )
     assert tester.execute("exit-code") == 42
     assert tester.execute("return-code") == 42
+
+
+@pytest.mark.parametrize(
+    "installed_script", [False, True], ids=["not installed", "installed"]
+)
+def test_run_project_script_sys_argv0(
+    installed_script: bool,
+    mocker: MockerFixture,
+    tmp_venv: VirtualEnv,
+    command_tester_factory: CommandTesterFactory,
+):
+    """
+    If RunCommand calls an installed script defined in pyproject.toml, sys.argv[0]
+    must be set to the full path of the script.
+    """
+    cli_script = "foo"
+
+    if installed_script:
+        if WINDOWS:
+            cli_script += ".exe"
+        cli_script = tmp_venv._bin_dir / cli_script
+        cli_script.touch()
+        if WINDOWS:
+            cli_script = str(cli_script).replace("\\", "\\\\")
+
+    poetry = Factory().create_poetry(
+        Path(__file__).parent.parent.parent / "fixtures" / "simple_project"
+    )
+    env_execute_mock = mocker.patch("poetry.utils.env.Env.execute", return_value=0)
+    tester = command_tester_factory("run", poetry, environment=tmp_venv)
+
+    tester.execute("foo status")
+
+    expected_sys_argv = f"sys.argv = ['{cli_script}', 'status']"
+    assert expected_sys_argv in env_execute_mock.call_args_list[0][0][2]
```

Comment:
This is a little magic for my liking:
* is more than one call expected?
* what are the other arguments?

How about `assert mock.call_args.args == [ANY, ANY, expected]` instead ?

Reply:
Another option is `m.assert_called_once_with([ANY, ANY, exp])`

Reply:
Oh, I missed the substring match... 
I guess that invalidates my suggestions but also shows how current version confused other developers.

Reply:
I feel the same. For that, I used once the following nice test lib https://callee.readthedocs.io

With it, I could rewrite the assert as
```
m.assert_called_once_with(ANY, ANY, Contains(expected_sys_argv))
```

Anyway, I will think about on how to improve this piece.

Reply:
@dimaqq I improved the test a bit

- Now I am using `mock.call_args[0][2]` to get the argument. It's better than `mock.call_args_list[0][0][2]`. Unfortunately I cannot use `mock.call_args.args[2]`, because `call_args.args` is available from Python 3.8, but poetry still requires Python 3.7 compatibility.
- I extracted the expression to a variable and added a comment explaining what is this.

Reply:
I rewrote the test using `assert_called_once_with`, it becomes better, I admit :smile: